src/bridge/connection.rs

```diff
@@ -0,0 +1,109 @@
+//! This module contains adaptations of the functions found in
+//! https://github.com/KillTheMule/nvim-rs/blob/master/src/create/tokio.rs
+
+use std::{
+    io::{self, Error, ErrorKind},
+    process::Stdio,
+};
+
+use nvim_rs::{error::LoopError, neovim::Neovim, Handler};
+use tokio::{
+    io::{split, AsyncRead, AsyncWrite},
+    net::{self, TcpStream},
+    process::Command,
+    spawn,
+    task::JoinHandle,
+};
+use tokio_util::compat::{TokioAsyncReadCompatExt, TokioAsyncWriteCompatExt};
+
+pub type ConnectionResult<W> = io::Result<(Neovim<W>, JoinHandle<Result<(), Box<LoopError>>>)>;
+pub type NeovimWriter = Box<dyn futures::AsyncWrite + Send + Unpin + 'static>;
+
+/// Connects to an existing Neovim instance.
+///
+/// Interprets `address` in the same way as `:help --server`: If it contains a `:` it's interpreted
+/// as a TCP/IPv4/IPv6 address. Otherwise it's interpreted as a named pipe or Unix domain socket
+/// path.
+pub async fn connect<H: Handler<Writer = NeovimWriter>>(
+    address: String,
+    handler: H,
+) -> io::Result<(Neovim<H::Writer>, JoinHandle<Result<(), Box<LoopError>>>)> {
+    if address.contains(":") {
+        connect_stream(TcpStream::connect(address).await?, handler).await
+    } else {
+        connect_ipc_socket(address, handler).await
+    }
+}
+
+/// Spawns and connects to an embedded Neovim instance.
+///
+/// stdin/stdout will be rewritten to `Stdio::piped()`
+pub async fn embed<H: Handler<Writer = NeovimWriter>>(
+    cmd: &mut Command,
+    handler: H,
+) -> ConnectionResult<H::Writer> {
+    let mut child = cmd.stdin(Stdio::piped()).stdout(Stdio::piped()).spawn()?;
+    let reader = child
+        .stdout
+        .take()
+        .ok_or_else(|| Error::new(ErrorKind::Other, "Can't open stdout"))?;
+    let writer = child
+        .stdin
+        .take()
+        .ok_or_else(|| Error::new(ErrorKind::Other, "Can't open stdin"))?;
+
+    connect_neovim(reader, writer, handler).await
+}
+
+async fn connect_stream<H, S>(stream: S, handler: H) -> ConnectionResult<H::Writer>
+where
+    S: AsyncRead + AsyncWrite + Send + Unpin + 'static,
+    H: Handler<Writer = NeovimWriter>,
+{
+    let (reader, writer) = split(stream);
+    connect_neovim(reader, writer, handler).await
+}
+
+async fn connect_neovim<H, R, W>(reader: R, writer: W, handler: H) -> ConnectionResult<H::Writer>
```

Comment:
Not quite sure if the naming causes more confusion than it helps. Currently, there are

- `connect` acting as kind of an entry-point into `connect_stream`/`connect_ipc_socket`
- `connect_stream` which does some setup work and forwards to `connect_neovim`
- `connect_ipc_socket` which sets up the right channel per platform and forwards to `connect_stream`
- `connect_neovim` carrying out the actual connection action

Maybe a wrapper holding a `NeovimWriter` and adding some construction methods could help clear this up. Or some sprinkled doc-comments. On the other hand, I did figure this out by myself (?).

Reply:
You're right, this is a mess. I've added some commits that restructure the code by introducing `NeovimInstance` and `NeovimSession` abstractions. Please let me know what you think.

The fixes for the comments above are in a separate commit, in case we don't want to go with the restructuring.

Reply:
The new solution is insanely clean, it looks great! Thanks!

Reply:
My pleasure!